---
title: "AllMines_Graph"
author: "Eudora Miao"
date: "4/10/2019"
output: html_document
---

# Importing data and pre-processing

```{r, warning=FALSE, error=FALSE, message=FALSE}
library(phyloseq)
library(ggplot2)
library(tidyverse)
library(RColorBrewer)
library(cowplot)
library(metacoder)
library(dplyr)
library(grDevices)

all_data0 <- import_biom(BIOMfilename = "all_otu_table_mapping.tax.sample.biom")
all_data0
```

## Checking sequencing depth

```{r}
# Make a data frame with a column for the read counts of each sample
sample_sum_df <- data.frame(sum = sample_sums(all_data0))
# Histogram of sample read counts
ggplot(sample_sum_df, aes(x = sum)) + 
  geom_histogram(color = "black", fill = "indianred", binwidth = 2500) +
  ggtitle("Distribution of sample sequencing depth") + 
  xlab("Read counts") +
  theme_gray() + theme(axis.title.y = element_blank())
```

## Standardizing by sequencing depth

```{r}
#Standardize abundances to the median sequencing depth
total <- median(sample_sums(all_data0))
standf <- function(x, t=total) round(t * (x/sum(x)))
all_data0.std <- transform_sample_counts(all_data0, standf)
#Filter taxa with cutoff 3.0 Coefficient of Variation
#BGM_data0.stdf <- filter_taxa(BGM_data0.std, function(x) sd(x)/mean(x) > 3.0, TRUE)
```

## Filtering "no hits"

```{r}
#Renaming taxonomy levels on tax_table
colnames(tax_table(all_data0.std)) <- c("Kingdom", "Phylum", "Class","Order", "Family", "Genus", "Species")
#Filtering no hit at Kingdom level
all_data <- subset_taxa(all_data0.std, Kingdom != "No blast hit")
```

# Taxa abudance at different rank levels

```{r, fig.width= 12, fig.height=6, fig.align='center'}
#summarizing by tax rank
all.class <- tax_glom(all_data, "Class")
plot_bar(all.class, fill = "Class", x = "Mine") + theme_gray()
```

## Subsetting datasets by study
```{r}
#Silvehill dataset
SH_data <- subset_samples(all_data, grepl("SH", sample_data(all_data)$Mine))
SH_field <- subset_samples(SH_data, grepl("Field", sample_data(SH_data)$Collection))
#BGM dataset
BGM_data <- subset_samples(all_data, grepl("BGM", sample_data(all_data)$Mine))
BGM_field <- subset_samples(BGM_data, grepl("Field", sample_data(BGM_data)$Collection))
#HKM dataset
HKM_data <- subset_samples(all_data, grepl("HKM", sample_data(all_data)$Mine))
#RGM dataset
RGM_data <- subset_samples(all_data, grepl("RGM", sample_data(all_data)$Mine))
#Mines (four mines field data combined)
Mines_data <- subset_samples(all_data, grepl("Field", sample_data(all_data)$Collection))
```

## Different mines field samples analyzed together
## Relative abudance of genus and class in different mine soil and root samples

```{r, fig.width= 12, fig.height=6, fig.align='center'}
TopNOTUs <- function(sample,N) {
  names(sort(taxa_sums(sample), TRUE)[1:N])
}
#Creating data frame from phyloseq object
top.Mines <- TopNOTUs(Mines_data, 100)
Mines.df <- prune_taxa(top.Mines, Mines_data) %>% psmelt()
#Sumarize data by site 
#test <- BGM.df %>% group_by(Genus) %>% summarise(sum(Abundance))
#plot by relative abundance
#Plot
pal <- colorRampPalette((brewer.pal(10, "Paired")))
Genus_soil <- ggplot(data = filter(Mines.df, Seq == "Soil"), aes(Site, Abundance, fill = Genus)) +
  geom_bar(stat = "identity", position = position_fill()) + coord_flip() +
    scale_fill_manual(values = (rev(pal(24)))) + 
    guides(fill = guide_legend(reverse = TRUE, ncol = 1, keyheight = 0.8)) +
    facet_grid(~ Seq, drop = TRUE) +
   theme(text = element_text(size = 15)) + theme_gray()


Genus_root <- ggplot(data = filter(Mines.df, Seq == "Root"), aes(Site, Abundance, fill = Genus)) +
  geom_bar(stat = "identity", position = position_fill()) + coord_flip() +
    scale_fill_manual(values = rev(pal(24))) + 
    guides(fill = FALSE) +
    facet_grid(~ Seq, drop = TRUE) +
   theme(text = element_text(size = 15)) + theme_gray()

plot_grid(Genus_root, Genus_soil, align = "h", rel_widths = c(1,1.2))

Class_soil <-ggplot(data = filter(Mines.df, Seq == "Soil"), aes(Site, Abundance, fill = Class)) +
  geom_bar(stat = "identity", position = position_fill()) + coord_flip() +
  scale_fill_manual(values = (pal(24))) + 
  guides(fill = guide_legend(ncol = 1, keyheight = 0.8)) +
  facet_grid(~ Seq, drop = TRUE) +
  theme(text = element_text(size = 15)) + theme_gray()

Class_root <-ggplot(data = filter(Mines.df, Seq == "Root"), aes(Site, Abundance, fill = Class)) +
  geom_bar(stat = "identity", position = position_fill()) + coord_flip() +
  scale_fill_manual(values = (pal(24))) + 
  guides(fill = FALSE) +  
  facet_grid(~ Seq, drop = TRUE) +
  theme(text = element_text(size = 15)) + theme_gray()

plot_grid(Class_root, Class_soil, align = "h", rel_widths = c(1,1.65))
```

```{r}
##Trying to stack the genus that belongs to the same class close to each other; also mannually set color so that the genus that belongs to the same class will be grouped in the same group of color

##This causes a lot of NAs
Mines.df$Genus <-  ordered(Mines.df$Genus, levels = c("g__unidentified", "g__Umbelopsis", "g__Tricholoma",
                                                      "g__Scleroderma", "g__Russula", "g__Rhizopogon",
                                                      "g__Pisolithus", "g__Oidiodendron", "g__Mortierella",
                                                      "g__Meliniomycetes", "g__Leptosphaeria", "g__Elaphomyces",
                                                      "g__Cortinarius", "g__Cenococcum", "g__Amanita",
                                                      "g__Boletus"))

(Genus_soil <- ggplot(data = filter(Mines.df, Seq == "Soil" & !is.na(Genus))) +
  geom_bar(aes(Site, Abundance, fill = Genus), stat = "identity", position = position_fill()) + coord_flip() +
    scale_fill_manual(values = colorRampPalette(brewer.pal(6, "Set1"))(16)) + 
    guides(fill = guide_legend(reverse = TRUE, ncol = 1, keyheight = 0.8)) +
    facet_grid(~ Seq, drop = TRUE) + #This part doesn't make sense because you already filter by soil
    theme(text = element_text(size = 15)) + theme_gray())

#These colors get repeated
scale_fill_manual(values = c("#e5f5f9", "#99d8c9", "#2ca25f", "#fee8c8", "#fdbb84", "#e34a33", "#e7e1ef",
                                 "#c994c7", "#dd1c77","#e5f5f9", "#99d8c9", "#2ca25f", "#fee8c8", "#fdbb84",
                                 "#e34a33", "#e7e1ef", "#c994c7", "#dd1c77", "#e5f5f9", "#99d8c9", "#2ca25f",
                                 "#fee8c8", "#fdbb84", "#e34a33", "#e7e1ef", "#c994c7", "#dd1c77")) +
```


```{r}
#Creating a table modeled after Huang et al. (2012)
sample_sum.std <- data.frame (sum = sample_sums (all_data0.std))
sample_sum.std$Sample <- c(rownames (sample_sum.std))

top10.BGM <- TopNOTUs(BGM_field, 10)
top10.BGM.df <- prune_taxa(top10.BGM, BGM_field) %>% psmelt()
View (top10.BGM.df)


top10.BGM.wsum <- merge (top10.BGM.df, sample_sum.std, by.x = "Sample", by.y = "Sample")

top10.BGM.wsum$otu_rel_abd <- (top10.BGM.df$Abundance)/(top10.BGM.df$sum)

top10.BGM.df %>%
<<<<<<< HEAD
  group_by(OTU, Seq, Site) %>%
  summarize(site_abundance = sum (Abundance)) %>%
  summarize(relative_abundance = site_abundance / sample_sums ())
=======
  add_column(otu_rel_abd = Abundance/sample_sum.std[rownames(sample_sum.std)==Sample])%>%
  group_by (OTU, Seq, Site) %>%
  summarize (site_abundance = sum (Abundance)) %>%
  summarize (relative_abundance = site_abundance / sample_sums ())
>>>>>>> 1f1214fe76e063bf00a5d0663f6b3466cbc7b74d
### what I am trying to do here is to calculate the relative abundance of each of the otus on different site (and also separated by "Soil" and "Roots"--which is classified under "Seq"). I wanted to use the sample_sums function to give me the overall number of reads from each site. How should I do this?

```


## Metacoder analysis for different mine field soil samples

```{r, warning=FALSE, message=FALSE}
#New variable
sample_data(BGM_data)$site_type <-
  substr (is.integer(as.integer(substr(sample_data(BGM_data)$Site), -1, -1)))

#Top 100 OTUs
top100.BGM <- TopNOTUs(BGM_data, 100)
top100.BGM.ps <- prune_taxa(top100.BGM, BGM_data)
#Converting to metacoder
obj <- parse_phyloseq(top100.BGM.ps)
# Convert counts to proportions
obj$data$otu_table <- calc_obs_props(obj,
                                     data = "otu_table",
                                     cols =
obj$data$sample_data$sample_id)
# Calculate per-taxon proportions
obj$data$tax_table <- calc_taxon_abund(obj,
                                       data = "otu_table",
                                       cols = obj$data$sample_data$sample_id)
#Compare treatments
obj$data$diff_table <- compare_groups(obj,
                                          data = "tax_table",
                                          cols = obj$data$sample_data$sample_id,
                                          groups = obj$data$sample_data$site_type)
```

```{r, fig.width= 12, fig.height=12, fig.align='center'}
#Tree visual
set.seed(1)

#Simply looking at all mine sites as a whole; node size and color representing number of observations
heat_tree(taxa::filter_taxa(obj, taxon_names == "c__Agaricomycetes", subtaxa = TRUE),
          node_label = taxon_names, node_size = n_obs, node_color = n_obs)

#Comparing different mines

heat_tree(taxa::filter_taxa(obj, taxon_names == "c__Agaricomycetes", subtaxa = TRUE),
          node_label = taxon_names, node_size = n_obs, node_color = total, edge_color = n_samples)


Tree1 <- metacoder::heat_tree(taxa::filter_taxa(obj, taxon_names == "c__Agaricomycetes", subtaxa = TRUE),
          node_size = n_obs, 
          node_label = taxon_names,
          node_color = log2_median_ratio,
          node_color_range = c("#a6611a","#dfc27d","#bdbdbd","#80cdc1","#018571"), 
          node_color_trans = "linear",
          node_label_max = 120,
          node_color_interval = c(-1, 1),
          edge_color_interval = c(-1, 1),
          node_size_axis_label = "Number of OTUs",
          node_color_axis_label = "Log2 ratio median proportions",
          title = "Site Differences",
          title_size = 0.03,
          initial_layout = "reingold-tilford", layout = "davidson-harel")
Tree1
```

## Funguild analysis of BGM soil and root samples

```{r, echo=FALSE, warning=FALSE, message=FALSE}
#Reading function from Colin's site 
#https://github.com/colinaverill/fg_assign.r
source("fg_assign.R")
str_charc <- function(x) {gsub('[f,p,c,o,k,g,s]__', '', x)}
#Modifying table to assign guild
test <- data.frame(tax_table(Mines_data)) %>%
  dplyr::rename(kingdom = Kingdom, 
                phylum = Phylum, 
                class = Class, 
                order = Order, 
                family = Family, 
                genus = Genus, 
                species = Species) %>% 
  rownames_to_column( var = "OTU_ID") %>%
  mutate_all(.funs = str_charc) %>%
  column_to_rownames("OTU_ID")
                    
#Assigning guild
test.assign <- fg_assign(test) %>%
  rownames_to_column(var = "OTU")
```

```{r, warning=FALSE, message=FALSE}
#Assigning guild to the tax table

new_tax <- data.frame(tax_table(Mines_data)) %>% 
  rownames_to_column(var = "OTU") %>% 
  left_join(test.assign[,c("OTU","guild")], by = "OTU") %>% 
  select(OTU, Kingdom, Phylum, Class, Order, Family, guild, Genus, Species) %>% 
  column_to_rownames(var = "OTU")
  
new_tax.0 <- tax_table(as.matrix(new_tax))
tax_table(Mines_data) <- new_tax.0
```

```{r, fig.width= 12, fig.height=10, fig.align='center'}
#Plotting abudance by guild
#Creating data frame from phyloseq object
top.Mines <- TopNOTUs(Mines_data, 100)
Mines.df <- prune_taxa(top.Mines, Mines_data) %>% psmelt()
#Plot
pal <- colorRampPalette((brewer.pal(8, "Paired")))
Guild_soil <- ggplot(data = filter(Mines.df, Seq == "Soil" & !is.na(guild)), aes(Site, Abundance, fill = guild)) +
  geom_bar(stat = "identity", position = position_fill()) + coord_flip() +
  scale_fill_manual(values = pal(16)) + 
  guides(fill = guide_legend(ncol = 3, keyheight = 0.6)) +
  facet_grid(~ Seq, drop = TRUE) +
  theme(text = element_text(size=40), axis.text.x = element_text(size = 40), legend.text = element_text(size=40)) +
  theme_gray()
Guild_soil.0 <- Guild_soil + theme(legend.position = "none")
Guild_soil.1 <- Guild_soil + theme(legend.position = "bottom")

Guild_root <- ggplot(data = filter(Mines.df, Seq == "Root" & !is.na(guild)), aes(Site, Abundance, fill = guild)) +
  geom_bar(stat = "identity", position = position_fill()) + coord_flip() +
  scale_fill_manual(values = pal(16)) + 
  guides(fill = FALSE) +
  facet_grid(~ Seq, drop = TRUE) +
  theme(text = element_text(size=40), axis.text.x = element_text(size = 40), legend.text = element_text(size=40)) + 
  theme_gray()


legend <- get_legend(Guild_soil.1)
P <- plot_grid(Guild_root, Guild_soil.0, align = "h", rel_heights = c(0.7,0.7))
ggdraw() + draw_plot(P, 0, 0.2, 1, 0.8) + draw_plot(legend, 0, -0.35, 1, 1, 2)

## I am having issue getting making the font of the texts larger
```

```{r}
###PCoA (adapted from https://github.com/alejorojas2/Corn_Michigan_data/blob/master/R_analysis/Community_analysis.md)

TopNOTUs <- function(sample,N) {
  names(sort(taxa_sums(sample), TRUE)[1:N])
}
#Creating data frame from phyloseq object -- extracting the top 100 taxa
top100.Mines <- TopNOTUs(Mines_data, 100)
Mines.df <- prune_taxa(names(top100.Mines), Mines_data)

#Ordination analysis
Mines_ord <- ordinate(Mines.df, "PCoA", "bray")
Mines_ord_plot <- plot_ordination(Mines.df, Mines_ord, color = "Site", shape = "Mine")
colors <- c("#084594",
"#2171b5",
"#4292c6",
"#6baed6",
"#9ecae1",
"#4a1486",
"#6a51a3",
"#807dba",
"#9e9ac8",
"#bcbddc",
"#d94801",
"#f16913",
"#238b45",
"#41ab5d",
"#74c476"
)

shapes <- c (15, 16, 17, 18)
Mines_ord_plot2 <- Mines_ord_plot + geom_point(size = 5, alpha = 0.7) + 
  scale_colour_manual(values = colors) + scale_shape_manual (values = shapes) +
  labs(title = "PCoA Mines") +
  theme_gray(base_size = 18) + labs(color = "Site", shape = "Mine") +
  theme(legend.text = element_text(size = 17),
        legend.key.size = unit(0.7, "cm"))

#Environmental factor analysis
Mines_env <- sample_data (Mines_data)
Mines_env <- Mines_env [,c (1, 3, 4, 5, 6, 7, 8, 11, 12, 16, 17, 18, 19, 20, 21, 22, 24, 25, 27, 28, 29, 31, 33, 34, 35, 37)]
Mines_env [,1:26] <- sapply (Mines_env [,1:26], as.numeric)
str (Mines_env)

Mines_envfit <- envfit(Mines_ord$vectors, permutations = 999, na.rm = TRUE)

fit_data <- as.data.frame(scores(Mines_envfit, display = "vectors")) %>%
  add_rownames(var = "Env.var") %>%
  bind_cols(data.frame(Mines_envfit$vectors$r, Mines_envfit$vectors$pvals)) %>%
  #rename(R2 = Oom_env.vectors.r, P.value = Oom_env.vectors.pvals) %>%
  arrange(Mines_envfit.vectors.pvals)
#########Error in validObject(.Object) : invalid class “sample_data” object: Sample Data must have non-zero dimensions.


```


